### PR TITLE
Support text segment boundary anchors

### PR DESCRIPTION
**Pull request by natecook1000**
This enables the `\y` and `\Y` anchors in regex literals and `Anchor.textSegmentBoundary` in the DSL.

Note: This also includes `UnicodeScalar` conformance to `RegexProtocol`, which acts like Unicode scalar literals in regex literals.

### CODE REVIEW

**Comment by natecook1000** [comment]
@swift-ci Please test Linux platform

**Comment by natecook1000** [comment]
@swift-ci Please test Linux platform

File: Sources/_StringProcessing/RegexDSL/DSL.swift [hunk 1]
```diff
@@ -39,6 +39,14 @@ extension Character: RegexProtocol {
   }
 }
 
+extension UnicodeScalar: RegexProtocol {
+  public typealias Match = Substring
+
+  public var regex: Regex<Match> {
+    .init(ast: atom(.scalar(self)))
```
Comment (milseman) [inline Sources/_StringProcessing/RegexDSL/DSL.swift:46]: We need to stop constructing ASTs in the DSL. Can you use DSLTree, and if not, let me know what's missing?

**Review by milseman — COMMENTED** [review]
Quick pass over.

File: Sources/_StringProcessing/RegexDSL/DSL.swift (hunk 1 above)
Comment (natecook1000) [reply Sources/_StringProcessing/RegexDSL/DSL.swift:46]: Switched this and the other primitives to use DSLTree 👍🏻

**Comment by natecook1000** [comment]
@swift-ci Please test Linux platform

File: Sources/_StringProcessing/ByteCodeGen.swift [hunk 2]
```diff
@@ -96,12 +96,16 @@ extension Compiler.ByteCodeGen {
       }
 
     case .textSegment:
-      // This we should be able to do!
-      throw Unsupported(#"\y (text segment)"#)
+      builder.buildAssert { (input, pos, _) in
+        // TODO: Unicode word boundary algorithm?
```
Comment (milseman) [inline Sources/_StringProcessing/ByteCodeGen.swift:100]:
```suggestion
        // FIXME: Grapheme or word based on options
```

File: Sources/_StringProcessing/ByteCodeGen.swift [hunk 3]
```diff
@@ -96,12 +96,16 @@ extension Compiler.ByteCodeGen {
...
+        input.isOnGraphemeClusterBoundary(pos)
+      }
 
     case .notTextSegment:
-      // This we should be able to do!
-      throw Unsupported(#"\Y (not text segment)"#)
+      builder.buildAssert { (input, pos, _) in
+        // TODO: Unicode word boundary algorithm?
```
Comment (milseman) [inline Sources/_StringProcessing/ByteCodeGen.swift:106]:
```suggestion
        // FIXME: Grapheme or word based on options
```

**Review by milseman — APPROVED** [review]
LGTM

**Comment by milseman** [comment]
@swift-ci please test

**Comment by natecook1000** [comment]
@swift-ci Please test Linux platform

**Comment by natecook1000** [comment]
@swift-ci Please test Linux platform

**Pull request merged**